org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ProcessTaskManager.java

```diff
@@ -46,159 +51,131 @@ public class ProcessTaskManager {
-					continue;
-			}
+					try {
+						this.compiler.reportProgress(Messages.bind(Messages.compilation_processing,
+								new String(unitToProcess.getFileName())));
+						if (this.compiler.options.verbose)
+							this.compiler.out.println(Messages.bind(Messages.compilation_process,
+									new String[] { String.valueOf(index + 1), String.valueOf(this.compiler.totalUnits),
+											new String(unitToProcess.getFileName()) }));
+						try {
+							this.compiler.process(unitToProcess, index);
+						} catch (AbortCompilation abortCompilation) {
+							unitToProcess.cleanUp();
+							throw abortCompilation;
+						} catch (Error | RuntimeException uncheckedThrowable) {
+							throw new RuntimeException(
+									"Internal Error compiling " + new String(unitToProcess.getFileName()), //$NON-NLS-1$
+									uncheckedThrowable);
+						}
+					} finally {
+						// cleanup compilation unit result, but only if not annotation processed.
+						if (cleanup) {
+							unitToProcess.cleanUp();
+						}
+					}
 
-			try {
-				this.compiler.reportProgress(Messages.bind(Messages.compilation_processing, new String(this.unitToProcess.getFileName())));
-				if (this.compiler.options.verbose)
-					this.compiler.out.println(
-						Messages.bind(Messages.compilation_process,
-						new String[] {
-							String.valueOf(index + 1),
-							String.valueOf(this.compiler.totalUnits),
-							new String(this.unitToProcess.getFileName())
-						}));
-				try {
-					this.compiler.process(this.unitToProcess, index);
-				} catch (AbortCompilation keptCancelation) {
-					throw keptCancelation;
-				} catch (Error | RuntimeException e) {
-					throw new RuntimeException("Internal Error compiling " + new String(this.unitToProcess.getFileName()), e); //$NON-NLS-1$
+					addNextUnit(unitToProcess);
+				} catch (Error | RuntimeException uncheckedThrowable) {
+					this.units.clear(); // make sure there is room for a premature stop signal
+					addNextUnit(uncheckedThrowable);
+					return;
 				}
-			} finally {
-				// cleanup compilation unit result, but only if not annotation processed.
-				if (this.unitToProcess != null && cleanup)
-					this.unitToProcess.cleanUp();
 			}
-
-			addNextUnit(this.unitToProcess);
-		} catch (Error | RuntimeException e) {
-			synchronized (this) {
-				this.processingTask = null;
-				this.caughtException = e;
-			}
-			return;
+		} finally {
+			addNextUnit(STOP_SIGNAL);
 		}
 	}
-}
+	synchronized CompilationUnitDeclaration getUnitToProcess(){
+		return this.lastUnitToProcess;
+	}
 
-public void shutdown() {
-	try {
-		Future<?> t = null;
-		synchronized (this) {
-			t = this.processingTask;
-			if (t != null) {
-				// stop processing on error:
-				this.processingTask = null;
-				notifyAll();
-			}
-		}
-		if (t != null) {
-			t.get(250, TimeUnit.MILLISECONDS); // do not wait forever
+	@Override
+	public void close() {
+		try {
+			this.processingTask.get(250, TimeUnit.MILLISECONDS); // do not wait forever
```

Comment:
should this be `this.processingTask.cancel(true)`?

Reply:
I tested: both units.clear() and cancel(true) let addNextUnit() progress. However cancel would only work if not anything else in the progress catches the interrupt. So i added both.